File: cpp/src/mkdir.cpp
#include "types.hpp"

#include "util.hpp"

namespace {

int do_mkdir(filefs::detail::FileSystemState& ffs, const std::string_view lastname,
             const std::uint32_t start_blockindex, filefs::detail::Block& start_block,
             const std::uint32_t cur_blockindex, filefs::detail::Block& cur_block,
             const std::uint32_t /*stop_blockindex*/, const std::uint16_t offset) {
  using namespace filefs;

  if (ffs.tmp.state == 0U && !ffs.tmpstart(1U)) {
    return 1;
  }

  detail::Block new_block{};
  detail::Block block2{};

  if (offset < BlockSize) {
    const std::uint32_t new_blockindex = ffs.genblockindex();
    if (new_blockindex == 0U) {
      if (ffs.tmp.state == 1U) {
        ffs.tmpstop();
      }
      return 1;
    }
    new_block.fill(0U);
    int k = detail::BlockHead;
    new_block[static_cast<std::size_t>(k)] = 0U;
    ++k;
    new_block[static_cast<std::size_t>(k)] = static_cast<std::uint8_t>('.');
    k += static_cast<int>(BlockNameMaxSize);
    detail::u32_to_b4(new_blockindex, std::span<std::uint8_t, 4>(new_block.data() + k, 4));
    k += 4;
    detail::u32_to_b4(new_blockindex, std::span<std::uint8_t, 4>(new_block.data() + k, 4));
    k += 4;
    detail::u16_to_b2(static_cast<std::uint16_t>(4 + 4 + 4 + 25 + 25),
                      std::span<std::uint8_t, 2>(new_block.data() + k, 2));
    k += 2;

    new_block[static_cast<std::size_t>(k)] = 0U;
    ++k;
    new_block[static_cast<std::size_t>(k)] = static_cast<std::uint8_t>('.');
    new_block[static_cast<std::size_t>(k + 1)] = static_cast<std::uint8_t>('.');
    k += static_cast<int>(BlockNameMaxSize);
    detail::u32_to_b4(start_blockindex, std::span<std::uint8_t, 4>(new_block.data() + k, 4));
    if (!ffs.writeblock(new_blockindex, new_block)) {
      if (ffs.tmp.state == 1U) {
        ffs.tmpstop();
      }
      return 1;
    }

    k = static_cast<int>(offset);
    cur_block[static_cast<std::size_t>(k)] = 0U;
    ++k;
    std::fill_n(cur_block.data() + k, BlockNameMaxSize, 0U);
    std::memcpy(cur_block.data() + k, lastname.data(), lastname.size());
    k += static_cast<int>(BlockNameMaxSize);
    detail::u32_to_b4(new_blockindex, std::span<std::uint8_t, 4>(cur_block.data() + k, 4));
    k += 4 + 4 + 2;
    const std::uint16_t new_offset = static_cast<std::uint16_t>(k);
    std::array<std::uint8_t, 2> b2{};
    detail::u16_to_b2(new_offset, b2);

    if (cur_blockindex == start_blockindex) {
      std::copy(b2.begin(), b2.end(), cur_block.begin() + detail::BlockOffset);
      if (!ffs.writeblock(cur_blockindex, cur_block)) {
        if (ffs.tmp.state == 1U) {
          ffs.tmpstop();
        }
        return 1;
      }
      if (ffs.tmp.state == 1U && !ffs.commit()) {
        return 1;
      }
    } else {
      if (!ffs.writeblock(cur_blockindex, cur_block)) {
        if (ffs.tmp.state == 1U) {
          ffs.tmpstop();
        }
        return 1;
      }
      std::copy(b2.begin(), b2.end(), start_block.begin() + detail::BlockOffset);
      if (!ffs.writeblock(start_blockindex, start_block)) {
        if (ffs.tmp.state == 1U) {
          ffs.tmpstop();
        }
        return 1;
      }
      if (ffs.tmp.state == 1U && !ffs.commit()) {
        return 1;
      }
    }
    return 0;
  }

  const std::uint32_t new_blockindex = ffs.genblockindex();
  if (new_blockindex == 0U) {
    if (ffs.tmp.state == 1U) {
      ffs.tmpstop();
    }
    return 1;
  }
  const std::uint32_t blockindex2 = ffs.genblockindex();
  if (blockindex2 == 0U) {
    if (ffs.tmp.state == 1U) {
      ffs.tmpstop();
    }
    return 1;
  }

  block2.fill(0U);
  int k = 8;
  detail::u32_to_b4(cur_blockindex, std::span<std::uint8_t, 4>(block2.data() + k, 4));
  k += 4;
  block2[static_cast<std::size_t>(k)] = 0U;
  ++k;
  std::fill_n(block2.data() + k, BlockNameMaxSize, 0U);
  std::memcpy(block2.data() + k, lastname.data(), lastname.size());
  k += static_cast<int>(BlockNameMaxSize);
  detail::u32_to_b4(new_blockindex, std::span<std::uint8_t, 4>(block2.data() + k, 4));
  k += 4 + 4 + 2;
  const std::uint16_t new_offset = static_cast<std::uint16_t>(k);
  if (!ffs.writeblock(blockindex2, block2)) {
    if (ffs.tmp.state == 1U) {
      ffs.tmpstop();
    }
    return 1;
  }

  new_block.fill(0U);
  k = detail::BlockHead;
  new_block[static_cast<std::size_t>(k)] = 0U;
  ++k;
  new_block[static_cast<std::size_t>(k)] = static_cast<std::uint8_t>('.');
  k += static_cast<int>(BlockNameMaxSize);
  detail::u32_to_b4(new_blockindex, std::span<std::uint8_t, 4>(new_block.data() + k, 4));
  k += 4;
  detail::u32_to_b4(new_blockindex, std::span<std::uint8_t, 4>(new_block.data() + k, 4));
  k += 4;
  detail::u16_to_b2(static_cast<std::uint16_t>(4 + 4 + 4 + 25 + 25),
                    std::span<std::uint8_t, 2>(new_block.data() + k, 2));
  k += 2;
  new_block[static_cast<std::size_t>(k)] = 0U;
  ++k;
  new_block[static_cast<std::size_t>(k)] = static_cast<std::uint8_t>('.');
  new_block[static_cast<std::size_t>(k + 1)] = static_cast<std::uint8_t>('.');
  k += static_cast<int>(BlockNameMaxSize);
  detail::u32_to_b4(start_blockindex, std::span<std::uint8_t, 4>(new_block.data() + k, 4));
  if (!ffs.writeblock(new_blockindex, new_block)) {
    if (ffs.tmp.state == 1U) {
      ffs.tmpstop();
    }
    return 1;
  }

  std::array<std::uint8_t, 2> b2{};
  detail::u16_to_b2(new_offset, b2);
  detail::u32_to_b4(blockindex2, std::span<std::uint8_t, 4>(cur_block.data() + 4, 4));

  if (cur_blockindex == start_blockindex) {
    detail::u32_to_b4(blockindex2, std::span<std::uint8_t, 4>(cur_block.data() + detail::BlockStopBlockindex, 4));
    std::copy(b2.begin(), b2.end(), cur_block.begin() + detail::BlockOffset);
    if (!ffs.writeblock(cur_blockindex, cur_block)) {
      if (ffs.tmp.state == 1U) {
        ffs.tmpstop();
      }
      return 1;
    }
    if (ffs.tmp.state == 1U && !ffs.commit()) {
      return 1;
    }
  } else {
    if (!ffs.writeblock(cur_blockindex, cur_block)) {
      if (ffs.tmp.state == 1U) {
        ffs.tmpstop();
      }
      return 1;
    }
    detail::u32_to_b4(blockindex2, std::span<std::uint8_t, 4>(start_block.data() + detail::BlockStopBlockindex, 4));
    std::copy(b2.begin(), b2.end(), start_block.begin() + detail::BlockOffset);
    if (!ffs.writeblock(start_blockindex, start_block)) {
      if (ffs.tmp.state == 1U) {
        ffs.tmpstop();
      }
      return 1;
    }
    if (ffs.tmp.state == 1U && !ffs.commit()) {
      return 1;
    }
  }

  return 0;
}

}  // namespace

namespace filefs {

int FileSystem::mkdir(const std::string_view pathname) {
  if (!state_ || !state_->mounted() || pathname.empty()) {
    return 1;
  }

  std::uint32_t blockindex = 0U;
  int start = 0;
  if (pathname.front() == '/') {
    blockindex = 1U;
    start = 1;
  } else if (state_->tmp.state == 0U) {
    blockindex = state_->pwd_blockindex;
  } else {
    blockindex = state_->tmp.pwd_blockindex;
  }

  std::array<char, BlockNameMaxSize + 2> s{};
  int slen = 0;
  for (int i = start; i < static_cast<int>(pathname.size()); ++i) {
    if (pathname[static_cast<std::size_t>(i)] == '/') {
      if (slen == 0) {
        continue;
      }
      const std::uint32_t index =
          state_->find_path_blockindex(blockindex, std::string_view{s.data(), static_cast<std::size_t>(slen)});
      if (index < 1U) {
        if (i == static_cast<int>(pathname.size()) - 1) {
          break;
        }
        return 1;
      }
      blockindex = index;
      slen = 0;
      continue;
    }
    s[static_cast<std::size_t>(slen++)] = pathname[static_cast<std::size_t>(i)];
    if (slen > static_cast<int>(BlockNameMaxSize)) {
      return 2;
    }
  }
  if (slen == 0) {
    return 3;
  }
  const std::string_view lastname{s.data(), static_cast<std::size_t>(slen)};
  if (lastname.size() > BlockNameMaxSize) {
    return 2;
  }

  detail::Block start_block{};
  detail::Block block{};
  if (!state_->readblock(blockindex, block)) {
    return 1;
  }
  start_block = block;
  const std::uint32_t start_blockindex = blockindex;

  const std::uint32_t stop_blockindex =
      detail::b4_to_u32(std::span<const std::uint8_t, 4>(block.data() + 12 + 1 + 14 + 4, 4));
  const std::uint16_t offset =
      detail::b2_to_u16(std::span<const std::uint8_t, 2>(block.data() + 12 + 1 + 14 + 4 + 4, 2));

  bool flag = false;
  std::uint32_t index = start_blockindex;
  for (;;) {
    int k = detail::BlockHead;
    for (int i = 0; i < detail::BlockItemMaxCount; ++i) {
      const std::uint8_t state = block[static_cast<std::size_t>(k)];
      ++k;
      if (detail::fixed_cstr(std::span<const std::uint8_t>(block.data() + k, BlockNameMaxSize)) != lastname) {
        k += static_cast<int>(BlockNameMaxSize) + 10;
        if (index == stop_blockindex && k + 1 >= static_cast<int>(offset)) {
          flag = true;
          break;
        }
        continue;
      }
      if ((state & 0x01U) == 0U) {
        return 3;
      }
      return 4;
    }
    if (flag) {
      break;
    }
    index = detail::b4_to_u32(std::span<const std::uint8_t, 4>(block.data() + 4, 4));
    if (index == 0U || !state_->readblock(index, block)) {
      return 1;
    }
  }

  return do_mkdir(*state_, lastname, start_blockindex, start_block, index, block, stop_blockindex, offset);
}

}  // namespace filefs

File: zig/src/mkdir.zig
const std = @import("std");
const types = @import("types.zig");
const util = @import("util.zig");
const txn = @import("txn.zig");
const block_mod = @import("block.zig");

pub fn doMkdir(self: anytype, lastname: []const u8, start_blockindex: u32, start_block: []u8, cur_blockindex: u32, cur_block: []u8, stop_blockindex: u32, offset: u16) i32 {
    _ = stop_blockindex;
    if (self.tmp.state == 0 and !txn.tmpstart(self, 1)) {
        return 1;
    }

    var b4: [4]u8 = undefined;
    var b2: [2]u8 = undefined;
    var new_block = util.blockZero();
    var block2 = util.blockZero();

    if (util.u16ToUsize(offset) < types.BlockSize) {
        const new_blockindex = block_mod.genblockindex(self);
        if (new_blockindex == 0) {
            if (self.tmp.state == 1) txn.tmpstop(self);
            return 1;
        }
        var k: usize = types.BlockHead;
        new_block[k] = 0;
        k += 1;
        new_block[k] = '.';
        k += types.BlockNameMaxSize;
        util.u32ToB4(new_blockindex, &b4);
        @memcpy(new_block[k .. k + 4], b4[0..]);
        k += 4;
        @memcpy(new_block[k .. k + 4], b4[0..]);
        k += 4;
        util.u16ToB2(4 + 4 + 4 + 25 + 25, &b2);
        @memcpy(new_block[k .. k + 2], b2[0..]);
        k += 2;
        new_block[k] = 0;
        k += 1;
        new_block[k] = '.';
        new_block[k + 1] = '.';
        k += types.BlockNameMaxSize;
        util.u32ToB4(start_blockindex, &b4);
        @memcpy(new_block[k .. k + 4], b4[0..]);
        if (!block_mod.writeblock(self, new_blockindex, new_block[0..])) {
            if (self.tmp.state == 1) txn.tmpstop(self);
            return 1;
        }

        k = util.u16ToUsize(offset);
        cur_block[k] = 0;
        k += 1;
        util.copyNameInto(cur_block[k .. k + types.BlockNameMaxSize], lastname);
        k += types.BlockNameMaxSize;
        util.u32ToB4(new_blockindex, &b4);
        @memcpy(cur_block[k .. k + 4], b4[0..]);
        k += 4;
        k += 4;
        k += 2;
        const new_offset = util.usizeToU16(k);
        util.u16ToB2(new_offset, &b2);

        if (cur_blockindex == start_blockindex) {
            @memcpy(cur_block[types.BlockOffset .. types.BlockOffset + 2], b2[0..]);
            if (!block_mod.writeblock(self, cur_blockindex, cur_block)) {
                if (self.tmp.state == 1) txn.tmpstop(self);
                return 1;
            }
            if (self.tmp.state == 1 and !txn.commit(self)) return 1;
        } else {
            if (!block_mod.writeblock(self, cur_blockindex, cur_block)) {
                if (self.tmp.state == 1) txn.tmpstop(self);
                return 1;
            }
            @memcpy(start_block[types.BlockOffset .. types.BlockOffset + 2], b2[0..]);
            if (!block_mod.writeblock(self, start_blockindex, start_block)) {
                if (self.tmp.state == 1) txn.tmpstop(self);
                return 1;
            }
            if (self.tmp.state == 1 and !txn.commit(self)) return 1;
        }
        return 0;
    }

    const new_blockindex = block_mod.genblockindex(self);
    if (new_blockindex == 0) {
        if (self.tmp.state == 1) txn.tmpstop(self);
        return 1;
    }
    const blockindex2 = block_mod.genblockindex(self);
    if (blockindex2 == 0) {
        if (self.tmp.state == 1) txn.tmpstop(self);
        return 1;
    }

    var k: usize = 8;
    util.u32ToB4(cur_blockindex, &b4);
    @memcpy(block2[k .. k + 4], b4[0..]);
    k += 4;
    block2[k] = 0;
    k += 1;
    util.copyNameInto(block2[k .. k + types.BlockNameMaxSize], lastname);
    k += types.BlockNameMaxSize;
    util.u32ToB4(new_blockindex, &b4);
    @memcpy(block2[k .. k + 4], b4[0..]);
    k += 4;
    k += 4;
    k += 2;
    const new_offset = util.usizeToU16(k);
    if (!block_mod.writeblock(self, blockindex2, block2[0..])) {
        if (self.tmp.state == 1) txn.tmpstop(self);
        return 1;
    }

    new_block = util.blockZero();
    k = types.BlockHead;
    new_block[k] = 0;
    k += 1;
    new_block[k] = '.';
    k += types.BlockNameMaxSize;
    util.u32ToB4(new_blockindex, &b4);
    @memcpy(new_block[k .. k + 4], b4[0..]);
    k += 4;
    @memcpy(new_block[k .. k + 4], b4[0..]);
    k += 4;
    util.u16ToB2(4 + 4 + 4 + 25 + 25, &b2);
    @memcpy(new_block[k .. k + 2], b2[0..]);
    k += 2;
    new_block[k] = 0;
    k += 1;
    new_block[k] = '.';
    new_block[k + 1] = '.';
    k += types.BlockNameMaxSize;
    util.u32ToB4(start_blockindex, &b4);
    @memcpy(new_block[k .. k + 4], b4[0..]);
    if (!block_mod.writeblock(self, new_blockindex, new_block[0..])) {
        if (self.tmp.state == 1) txn.tmpstop(self);
        return 1;
    }

    util.u16ToB2(new_offset, &b2);
    util.u32ToB4(blockindex2, &b4);
    @memcpy(cur_block[4..8], b4[0..]);
    if (cur_blockindex == start_blockindex) {
        @memcpy(cur_block[types.BlockStopBlockindex .. types.BlockStopBlockindex + 4], b4[0..]);
        @memcpy(cur_block[types.BlockOffset .. types.BlockOffset + 2], b2[0..]);
        if (!block_mod.writeblock(self, cur_blockindex, cur_block)) {
            if (self.tmp.state == 1) txn.tmpstop(self);
            return 1;
        }
        if (self.tmp.state == 1 and !txn.commit(self)) return 1;
    } else {
        if (!block_mod.writeblock(self, cur_blockindex, cur_block)) {
            if (self.tmp.state == 1) txn.tmpstop(self);
            return 1;
        }
        @memcpy(start_block[types.BlockStopBlockindex .. types.BlockStopBlockindex + 4], b4[0..]);
        @memcpy(start_block[types.BlockOffset .. types.BlockOffset + 2], b2[0..]);
        if (!block_mod.writeblock(self, start_blockindex, start_block)) {
            if (self.tmp.state == 1) txn.tmpstop(self);
            return 1;
        }
        if (self.tmp.state == 1 and !txn.commit(self)) return 1;
    }
    return 0;
}

pub fn mkdir(self: anytype, pathname: []const u8) i32 {
    if (pathname.len == 0 or self.fp == null) return 1;
    var blockindex: u32 = undefined;
    var start: usize = undefined;
    if (pathname[0] == '/') {
        blockindex = 1;
        start = 1;
    } else {
        blockindex = util.currentPwdBlockindex(self);
        start = 0;
    }

    var s: [types.BlockNameMaxSize + 2]u8 = [_]u8{0} ** (types.BlockNameMaxSize + 2);
    var slen: usize = 0;
    var i: usize = start;
    while (i < pathname.len) : (i += 1) {
        if (pathname[i] == '/') {
            if (slen == 0) continue;
            const index = block_mod.findPathBlockindex(self, blockindex, s[0..slen]);
            if (index < 1) {
                if (i == pathname.len - 1) break;
                return 1;
            }
            blockindex = index;
            slen = 0;
            continue;
        }
        s[slen] = pathname[i];
        slen += 1;
        if (slen > types.BlockNameMaxSize) return 2;
    }
    if (slen == 0) return 3;
    var lastname_buf: [types.BlockNameMaxSize + 1]u8 = [_]u8{0} ** (types.BlockNameMaxSize + 1);
    @memcpy(lastname_buf[0..slen], s[0..slen]);
    const lastname = lastname_buf[0..slen];
    if (lastname.len > types.BlockNameMaxSize) return 2;

    var start_block = util.blockZero();
    var block = util.blockZero();
    var b4: [4]u8 = undefined;
    var b2: [2]u8 = undefined;
    if (!block_mod.readblock(self, blockindex, block[0..])) return 1;
    @memcpy(start_block[0..], block[0..]);
    const start_blockindex = blockindex;

    @memcpy(b4[0..], block[12 + 1 + 14 + 4 .. 12 + 1 + 14 + 8]);
    const stop_blockindex = util.b4ToU32(b4[0..]);
    @memcpy(b2[0..], block[12 + 1 + 14 + 4 + 4 .. 12 + 1 + 14 + 4 + 6]);
    const offset = util.b2ToU16(b2[0..]);

    var flag = false;
    var index = start_blockindex;
    while (true) {
        var k: usize = types.BlockHead;
        var item: usize = 0;
        while (item < types.BlockItemMaxCount) : (item += 1) {
            const state = block[k];
            k += 1;
            @memset(s[0..], 0);
            @memcpy(s[0..types.BlockNameMaxSize], block[k .. k + types.BlockNameMaxSize]);
            k += types.BlockNameMaxSize;
            if (!std.mem.eql(u8, util.fixedCStr(s[0..types.BlockNameMaxSize + 1]), lastname)) {
                k += 10;
                if (index == stop_blockindex and k + 1 >= util.u16ToUsize(offset)) {
                    flag = true;
                    break;
                }
                continue;
            }
            if ((state & 0x01) == 0) return 3;
            return 4;
        }
        if (flag) break;
        @memcpy(b4[0..], block[4..8]);
        index = util.b4ToU32(b4[0..]);
        if (index == 0) return 1;
        if (!block_mod.readblock(self, index, block[0..])) return 1;
    }

    return doMkdir(self, lastname, start_blockindex, start_block[0..], index, block[0..], stop_blockindex, offset);
}
